### PR TITLE
sort! rules

### PR DESCRIPTION
**Pull request by jgreener64**
I made Enzyme rules for `sort!` as suggested in #880. These are my first Enzyme rules so will need some checking.

I put them in `src/internal_rules.jl` for now, let me know if they should go somewhere else. I have tests too which I can add later.

### CODE REVIEW

File: src/internal_rules.jl [hunk 1]
```diff
@@ -99,3 +99,56 @@ end
...
 end
+
+function EnzymeRules.forward(
+        ::Const{typeof(sort!)},
+        RT::Type{<:Union{Const, DuplicatedNoNeed, Duplicated}},
+        xs::Duplicated;
+        kwargs...
+    )
+    inds = sortperm(xs.val; kwargs...)
+    xs.val .= xs.val[inds]
```
Comment (wsmoses) [inline src/internal_rules.jl:110]: Looks like sortperm itself just makes a list of 1:N, then sorts that in place using the original data as the order: https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1756

Could we change this to first sort the derivative array in place using the primal array as the order, then do the actual sort on the primal? That way we don't have that temporary

File: src/internal_rules.jl [hunk 2]
```diff
@@ -99,3 +99,56 @@ end
...
+    xs.dval .= xs.dval[inds]
+    if RT <: Const
+        return xs.val
```
Comment (wsmoses) [inline src/internal_rules.jl:113]: Looks good, but mind also adding support for BatchDuplicated/BatchDuplicatedNoNeed?

also tests?

File: src/internal_rules.jl [hunk 3]
```diff
@@ -99,3 +99,56 @@ end
...
+
+function EnzymeRules.augmented_primal(
+        config::EnzymeRules.ConfigWidth{1},
+        ::Const{typeof(sort!)},
+        RT::Type{<:Union{Const, DuplicatedNoNeed, Duplicated}},
+        xs::Duplicated;
+        kwargs...
+    )
+    inds = sortperm(xs.val; kwargs...)
+    if EnzymeRules.needs_primal(config)
```
Comment (wsmoses) [inline src/internal_rules.jl:129]: we still need to sort the primal and shadow here, like in the forward pass. The only difference is that we should now also cache the permutation, and the reverse pass will invert the permutation

**Comment by jgreener64** [comment]
Added tests and changed the augmented primal. I will address the other points later.

**Comment by jgreener64** [comment]
Added a rule for batched mode.

I had a look at sorting without allocating the index array but couldn't get it to work. The `Perm` order at https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1756 is specifically for ordering indices, rather than generic ordering of another array.

File: test/runtests.jl [hunk 4]
```diff
@@ -77,6 +77,7 @@ include("typetree.jl")
     include("rrules.jl")
     include("kwrules.jl")
     include("kwrrules.jl")
+    include("internal_rules.jl")
```
Comment (wsmoses) [inline test/runtests.jl:80]: Maybe also move the A \ B rule here (its at the end of the file).

**Review by wsmoses — APPROVED** [review]

**Comment by jgreener64** [comment]
Is this ready for merge? Test failures look unrelated.

**Comment by matinraayai** [comment]
@jgreener64 will this rule also cover out-of-place ```sort``` across different dims?

**Comment by jgreener64** [comment]
In general `sort` will call this rule yes since it calls `sort!` internally: https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1495-L1500.

`sort` with `dims` though maybe not as that goes down a different path: https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1817. Give it a go and see what happens?

**Comment by matinraayai** [comment]
> In general `sort` will call this rule yes since it calls `sort!` internally: https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1495-L1500.
> 
> `sort` with `dims` though maybe not as that goes down a different path: https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1817. Give it a go and see what happens?

This is what I'm getting on Enzyme Master right now:
```
Illegal updateAnalysis prev:{[-1]:Pointer, [-1,-1]:Float@double} new: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}
val:   %47 = bitcast double addrspace(13)* %46 to i64 addrspace(13)*, !dbg !291 origin=  store i64 %42, i64 addrspace(13)* %47, align 8, !dbg !291, !tbaa !207, !alias.scope !210, !noalias !211

Caused by:
Stacktrace:
 [1] setindex!
   @ ./array.jl:1019
 [2] _setindex_ra!
   @ ./reinterpretarray.jl:537
 [3] setindex!
   @ ./reinterpretarray.jl:505
 [4] _sort!
   @ ./sort.jl:906
```
I was hoping this PR would fix this issue, but if not maybe I can help add something new to the rule to cover the ```dim``` keyword?

**Comment by jgreener64** [comment]
Does that error occur with this PR as well?

**Comment by jgreener64** [comment]
It looks like a specific rule needs to be written for this path: https://github.com/JuliaLang/julia/blob/750df9fb5bede16f321f5d5405943d12aec7b83e/base/sort.jl#L1817.

Though as a short-term fix you can copy the array and call `sort!`, that should hit the above rules and be fast.

**Comment by matinraayai** [comment]
@jgreener64 I will do that (I don't care about the unsorted array). I thought the rules would apply to the `!` variant too however?

**Comment by jgreener64** [comment]
The rules in this PR will always apply to `sort!` and will apply to `sort` in cases where `dims` is not set. So the workaround for `sort` with `dims` is to copy the array and use `sort!`, which always hits these rules.

**Pull request merged**